Add sort by option to node selector

bekriebel opened the pull request:
Allow the final decision of a selected node to be based on different
sorting options rather than just random.

Add a `sysload` sorting option to select the lowest loaded node when
multiple are suitable for the chosen selector.

Add tests for sysload sort_by option

Closes: #598

davidzhao commented on pkg/routing/selector/cpuload.go:
```diff
@@ -31,12 +29,11 @@ func (s *CPULoadSelector) filterNodes(nodes []*livekit.Node) ([]*livekit.Node, e
 }
 
-func (s *CPULoadSelector) SelectNode(nodes []*livekit.Node) (*livekit.Node, error) {
+func (s *CPULoadSelector) SelectNode(nodes []*livekit.Node, sortBy string) (*livekit.Node, error) {
```
instead of changing `SelectNode`, IMO it's cleaner to have sortby as a field in the node selector itself during construction.

bekriebel replied:
I can do that if you prefer. I thought this seemed cleaner since it's really only needed in the SelectNode call, but I can do an update to move it to the selector constructor.

davidzhao replied:
that would be great 🙏  I'd like to keep the `SelectNode` interface so that it only needs to take in `nodes`. Especially if we started adding other options that are specific to certain selectors.

bekriebel commented on pkg/routing/selector/utils.go:
```diff
@@ -39,3 +49,23 @@ func LimitsReached(limitConfig config.LimitConfig, nodeStats *livekit.NodeStats)
+		  return GetNodeSysload(nodes[i]) < GetNodeSysload(nodes[j])
+		})
+		return nodes[0], nil
+	default:
```
> can we default to random, the previous default?

It gets defaulted in the config if unset here: https://github.com/livekit/livekit-server/pull/599/files#diff-a3d824da3c42420cd5cbb0a4a2c0e7b5bfddd819652788a0596d195dc6e31fa5R240. I could do it here as well, but it seemed better to catch the default in this way so it would throw an error if there was a typo in a config or some unknown value was set. Do you still want me to also set the default here to random?

davidzhao replied:
that makes sense, I agree this is better to catch potential typos 👍

bekriebel commented:
@davidzhao I updated the PR to switch the SortBy value to be in the constructors. I also went ahead and added some more options to sort by since that's fairly trivial on anything that there are already stats for.

davidzhao commented on pkg/routing/selector/interfaces.go:
```diff
@@ -24,20 +24,22 @@ func CreateNodeSelector(conf *config.Config) (NodeSelector, error) {
 		return s, nil
 	case "random":
-		return &RandomSelector{}, nil
+		return &RandomSelector{conf.NodeSelector.SortBy}, nil
```
if the `Random` selector is to take in a sort field that's not random, this would be pretty confusing as the name becomes meaningless. wdyt about leaving this one alone?

bekriebel replied:
I thought about this, but at the moment "random" is the only one that would expose all nodes without some kind of limit. Ideally, the "random" selector would actually be called something like "all" or "any", but I didn't want to change something that was the default.

I'm happy to remove this, as I agree it's a little confusing, but I could see a use case for it.

davidzhao replied:
how about let's do this: remove it for now to eliminate confusion. Then we can always add a new node selector for the use cases that desire it.

I don't think anyone should be using the random selector FWIW, just wanting to keep an eye on things that introduce increased cognitive load.

bekriebel replied:
Actually, what do you think about renaming the `random` selector to `any` and keeping `random` in interfaces pointing to `any` with a deprecation warning? Looks like that's a trivial code change that I could add here as a separate commit if it sounds good.

bekriebel replied:
I updated this PR to rename `random`. Personally, I prefer this approach. However, I opened #612 if you prefer to go with sort by not applying to the random selector,

davidzhao replied:
looks good to me!

davidzhao reviewed (approved):
looks good! so nice to be able to expose other sort options.
